core/local/local.go

```diff
@@ -23,527 +23,363 @@ package local
-			leftovers := time.Since(lastTick)
-			select {
-			case <-pause:
-				e.Logger.Debug("Local: No longer paused")
-				lastTick = time.Now().Add(-leftovers)
-			case <-ctx.Done():
-				e.Logger.Debug("Local: Terminated while in paused state")
-				return nil
-			}
-		}
-
-		// Dumb hack: we don't wanna start any more iterations than the max, but we can't
-		// conditionally select on a channel either...so, we cheat: swap out the flow channel for a
-		// nil channel (writing to nil always blocks) if we don't wanna write an iteration.
-		flow := vuFlow
-		end := atomic.LoadInt64(&e.endIters)
-		partials := atomic.LoadInt64(&e.partIters)
-		if end >= 0 && partials >= end {
-			flow = nil
-		}
-
+	initializedVUs := new(uint64)
+	vusFmt := pb.GetFixedLengthIntFormat(int64(vusToInitialize))
+	e.initProgress.Modify(
+		pb.WithProgress(func() (float64, string) {
+			doneVUs := atomic.LoadUint64(initializedVUs)
+			return float64(doneVUs) / float64(vusToInitialize),
+				fmt.Sprintf(vusFmt+"/%d VUs initialized", doneVUs, vusToInitialize)
+		}),
+	)
+
+	for vuNum := uint64(0); vuNum < vusToInitialize; vuNum++ {
 		select {
-		case flow <- partials:
-			// Start an iteration if there's a VU waiting. See also: the big comment block above.
-			atomic.AddInt64(&e.partIters, 1)
-		case t := <-ticker.C:
-			// Every tick, increment the clock, see if we passed the end point, and process stages.
-			// If the test ends this way, set a cutoff point; any samples collected past the cutoff
-			// point are excluded.
-			d := t.Sub(lastTick)
-			lastTick = t
-
-			end := time.Duration(atomic.LoadInt64(&e.endTime))
-			at := time.Duration(atomic.AddInt64(&e.time, int64(d)))
-			if end >= 0 && at >= end {
-				e.Logger.WithFields(log.Fields{"at": at, "end": end}).Debug("Local: Hit time limit")
-				cutoff = time.Now()
-				return nil
-			}
-
-			stages := e.stages
-			if len(stages) > 0 {
-				vus, keepRunning := ProcessStages(startVUs, stages, at)
-				if !keepRunning {
-					e.Logger.WithField("at", at).Debug("Local: Ran out of stages")
-					cutoff = time.Now()
-					return nil
-				}
-				if vus.Valid {
-					if err := e.SetVUs(vus.Int64); err != nil {
-						return err
-					}
-				}
-			}
-		case sampleContainer := <-vuOut:
-			engineOut <- sampleContainer
-		case <-iterDone:
-			// Every iteration ends with a write to iterDone. Check if we've hit the end point.
-			// If not, make sure to include an Iterations bump in the list!
-			var tags *stats.SampleTags
-			if e.Runner != nil {
-				tags = e.Runner.GetOptions().RunTags
-			}
-			engineOut <- stats.Sample{
-				Time:   time.Now(),
-				Metric: metrics.Iterations,
-				Value:  1,
-				Tags:   tags,
-			}
-
-			end := atomic.LoadInt64(&e.endIters)
-			at := atomic.AddInt64(&e.iters, 1)
-			if end >= 0 && at >= end {
-				e.Logger.WithFields(log.Fields{"at": at, "end": end}).Debug("Local: Hit iteration limit")
-				return nil
+		case err := <-doneInits:
+			if err != nil {
+				return err
```

Comment:
I am not sure it is worth it to actually wait for all errors and combine them and then return them. 
Maybe have a comment if we won't do it

Reply:
I don't think combining them is practical, because after we encounter this first error, we abort the initialization of the remaining VUs by cancelling the context. So any errors they return are likely to be from that, which won't be very useful to users, it will just be noise that obscures the actual error cause.

Reply:
but yeah, I should add a comment that explains this

Reply:
Resolved by https://github.com/loadimpact/k6/pull/1007/commits/157c05547ed8b6d89bf06f6b11d2cbc0ba931f0b